subprojects/build-cache/src/integTest/groovy/org/gradle/caching/configuration/BuildCacheCompositeConfigurationIntegrationTest.groovy

```diff
@@ -106,10 +109,55 @@ class BuildCacheCompositeConfigurationIntegrationTest extends AbstractIntegratio
 
         pathToCacheDirMap == [
             ":": mainCache.cacheDir,
+            ":i1": mainCache.cacheDir,
+            ":i1:buildSrc": mainCache.cacheDir,
+            ":i2": mainCache.cacheDir,
             ":buildSrc": buildSrcCache.cacheDir
         ]
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/4216")
+    def "build cache service is closed only after all included builds are finished"() {
+        def localCache = new TestBuildCache(file("local-cache"))
+
+        buildTestFixture.withBuildInSubDir()
+        ['i1', 'i2'].each {
+            multiProjectBuild(it, ['first', 'second']) {
+                buildFile << """
+                gradle.startParameter.setTaskNames(['build'])
+                println gradle
+                allprojects {
+                    apply plugin: 'java-library'
+                }
+                """
+            }
+        }
+
+        settingsFile << localCache.localCacheConfiguration() << """
+            includeBuild "i1"
+            includeBuild "i2"
+        """
+        buildFile << """             
+            apply plugin: 'java'
+            println gradle
+            gradle.startParameter.setTaskNames(['build'])
+            processResources {
+                dependsOn gradle.includedBuild('i1').task(':processResources')
+                dependsOn gradle.includedBuild('i2').task(':processResources')
+            }
+        """
+
+        expect:
+        succeeds '--parallel'
```

Comment:
you don't need `--parallel` as the tasks for included builds are run in parallel anyway.

Reply:
Good catch, @adammurdoch. Removed.